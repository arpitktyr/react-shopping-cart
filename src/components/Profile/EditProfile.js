import React from "react";
import { useForm } from "react-hook-form";

const EditProfile = ({ userData, onSave }) => {
  const { register, handleSubmit, formState } = useForm({
    mode: "onBlur",
    defaultValues: {
      name: userData?.name,
      email: userData?.email,
      pincode: userData?.pincode,
      address: userData?.address,
      mobile: userData?.mobile,
    },
  });
  const { errors } = formState;

  const handleProfileForm = (data) => {
    const { name, email, address, pincode, mobile } = data;
    onSave({
      name,
      email,
      pincode,
      address,
      mobile,
    });
  };

  return (
    <form
      name="profileForm"
      onSubmit={handleSubmit(handleProfileForm)}
      style={{ width: "100%", display: "flex", flexDirection: "column" }}
    >
      <div className="form-group">
        <label htmlFor="name">Name</label>
        <input
          className="form-control"
          type="text"
          id="name"
          {...register("name", {
            required: {
              value: true,
              message: "Name is required.",
            },
            pattern: {
              value: /^[a-zA-Z ]+$/,
              message: "Invalid name.",
            },
          })}
          aria-invalid={errors.name ? "true" : "false"}
        />
        {errors.name && (
          <span className="text-danger">{errors.name?.message}</span>
        )}
      </div>
      <div className="form-group">
        <label htmlFor="email">Email</label>
        <input
          className="form-control"
          type="email"
          id="email"
          {...register("email", {
            required: {
              value: true,
              message: "Email is required.",
            },
            pattern: {
              value: /^[^\s@]+@[^\s@]+\.[^\s@]+$/,
              message: "Invalid email format.",
            },
          })}
          aria-invalid={errors.email ? "true" : "false"}
        />
        {errors.email && (
          <span className="text-danger">{errors.email?.message}</span>
        )}
      </div>
      <div className="form-group">
        <label htmlFor="pincode">Pincode</label>
        <input
          className="form-control"
          type="text"
          id="pincode"
          {...register("pincode", {
            required: {
              value: true,
              message: "Pincode is required.",
            },
            pattern: {
              value: /^[0-9]{6}$/,
              message: "Enter valid Pincode.",
            },
          })}
          aria-invalid={errors.pincode ? "true" : "false"}
        />{" "}
        {errors.pincode && (
          <span className="text-danger">{errors.pincode?.message}</span>
        )}
      </div>
      <div className="form-group">
        <label htmlFor="address">Address</label>
        <textarea
          className="form-control"
          id="address"
          {...register("address", {
            required: {
              value: true,
              message: "Address is required.",
            },
          })}
          aria-invalid={errors.address ? "true" : "false"}
        />
        {errors.address && (
          <span className="text-danger">{errors.address?.message}</span>
        )}
      </div>
      <div className="form-group">
        <label htmlFor="mobile">Mobile</label>
        <input
          className="form-control"
          type="text"
          id="mobile"
          {...register("mobile", {
            required: {
              value: true,
              message: "Mobile is required.",
            },
            pattern: {
              value: /^[0-9]{10}$/,
              message: "Enter valid Mobile No.",
            },
          })}
          aria-invalid={errors.mobile ? "true" : "false"}
        />
        {errors.mobile && (
          <span className="text-danger">{errors.mobile?.message}</span>
        )}
      </div>
      <button name="save" className="btn btn-success" type="submit">
        Save
      </button>
    </form>
  );
};

export default EditProfile;
