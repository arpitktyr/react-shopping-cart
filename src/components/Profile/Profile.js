import React, { useState, useContext, useEffect } from "react";
import EditProfile from "./EditProfile";
import { UserContext } from "../../context/user-context";

const Profile = () => {
  const { userName, userEmail } = useContext(UserContext);

  const [userData, setUserData] = useState({
    name: userName,
    email: userEmail,
    pincode: "209727",
    address: "Kanpur, Uttar Pradesh",
    mobile: "9956455678",
  });

  useEffect(() => {
    const data = { ...userData, email: userEmail, name: userName };
    setUserData(data);
  }, [userEmail, userName]);

  const handleSave = (newUserData) => {
    console.log(newUserData);
    setUserData(newUserData);
  };

  return (
    <div className="container my-4">
      <h1 className="text-center">Profile</h1>
      <p className="text-center">
        Keep your profile up-to-date to ensure accurate billing and shipping
        information for all your orders.
      </p>
      <div className="container">
        <div className="row justify-content-md-center">
          <div className="col col-sm-6">
            <EditProfile userData={userData} onSave={handleSave} />
          </div>
        </div>
      </div>
    </div>
  );
};
export default Profile;
