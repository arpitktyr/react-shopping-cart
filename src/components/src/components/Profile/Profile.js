
import React, {useState} from 'react'
import EditProfile from './EditProfile';

const Profile = ()=>{
    const [userData, setUserData] = useState({
        name: "Aman Katiyar",
        email: "aman@gmail.com",
        pincode: "209727",
        address: "Kanpur, Uttar Pradesh",
        mobile: "9956455678",
      });
    
      const handleSave = (newUserData) => {
        console.log(newUserData);
        setUserData(newUserData);
      };

 return (
    <div className="container my-4"> 
 <h1 className='text-center'>
    Profile
 </h1>
 <p className='text-center'>Keep your profile up-to-date to ensure accurate billing and shipping information for all your orders.</p>
 <div class="container">
  <div class="row justify-content-md-center">
    <div className='col col-sm-6' >
 <EditProfile userData={userData} onSave={handleSave}/>
 </div>
 </div>
 </div></div>
 );
}
export default Profile;