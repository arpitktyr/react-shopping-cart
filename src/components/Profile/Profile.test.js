import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Profile from "./Profile";
import EditProfile from "./EditProfile";
import { BrowserRouter } from "react-router-dom";
import UserContextProvider from "../../context/user-context";
import { waitFor } from "@testing-library/react";
import axios from "axios";
jest.mock("axios");

const userData = {
  name: "aman",
  email: "aman@gmail.com",
  pincode: "209727",
  address: "Kanpur, Uttar Pradesh",
  mobile: "9956455678",
};
const onSaveMock = jest.fn();

describe("EditProfile component", () => {
  it("User should be able to write in input type", async () => {
    render(
      <UserContextProvider>
        <BrowserRouter>
          <EditProfile userData={userData} onSave={onSaveMock} />
        </BrowserRouter>
      </UserContextProvider>
    );
    const nameInput = await screen.findByLabelText("Name");
    fireEvent.change(nameInput, { target: { value: "amanktyr" } });
    expect(nameInput.value).toBe("amanktyr");

    const emailInput = screen.getByLabelText("Email");
    fireEvent.change(emailInput, { target: { value: "amanktyr@gmaill.com" } });
    expect(emailInput.value).toBe("amanktyr@gmaill.com");

    const addInput = screen.getByLabelText("Address");
    fireEvent.change(addInput, { target: { value: "India" } });
    expect(addInput.value).toBe("India");

    const pincodeInput = screen.getByLabelText("Pincode");
    fireEvent.change(pincodeInput, { target: { value: "208024" } });
    expect(pincodeInput.value).toBe("208024");

    const mobileInput = screen.getByLabelText("Mobile");
    fireEvent.change(mobileInput, { target: { value: "2345678989" } });
    expect(mobileInput.value).toBe("2345678989");
  });

  it("should updated user data when the form is submitted", () => {
    render(
      <UserContextProvider>
        <BrowserRouter>
          <EditProfile userData={userData} onSave={onSaveMock} />
        </BrowserRouter>
      </UserContextProvider>
    );
    const nameInput = screen.getByLabelText("Name");
    const emailInput = screen.getByLabelText("Email");
    const pincodeInput = screen.getByLabelText("Pincode");
    const addressInput = screen.getByLabelText("Address");
    const mobileInput = screen.getByLabelText("Mobile");
    const saveButton = screen.getByText("Save");

    fireEvent.change(nameInput, { target: { value: "aman" } });
    fireEvent.change(emailInput, { target: { value: "aman@gmail.com" } });
    fireEvent.change(pincodeInput, { target: { value: "209727" } });
    fireEvent.change(addressInput, {
      target: { value: "Kanpur, Uttar Pradesh" },
    });
    fireEvent.change(mobileInput, { target: { value: "9956455678" } });
    fireEvent.click(saveButton);
    expect(nameInput.value).toBe("aman");
    expect(emailInput.value).toBe("aman@gmail.com");
    expect(pincodeInput.value).toBe("209727");
    expect(mobileInput.value).toBe("9956455678");
  });

  it("should call the onSave function", async () => {
    render(
      <UserContextProvider>
        <BrowserRouter>
          <EditProfile userData={userData} onSave={onSaveMock} />
        </BrowserRouter>
      </UserContextProvider>
    );
    const nameInput = screen.getByLabelText("Name");
    const emailInput = screen.getByLabelText("Email");
    const pincodeInput = screen.getByLabelText("Pincode");
    const addressInput = screen.getByLabelText("Address");
    const mobileInput = screen.getByLabelText("Mobile");
    const saveButton = screen.getByText("Save");

    fireEvent.change(nameInput, { target: { value: "aman" } });
    fireEvent.change(emailInput, { target: { value: "aman@gmail.com" } });
    fireEvent.change(pincodeInput, { target: { value: "209727" } });
    fireEvent.change(addressInput, {
      target: { value: "Kanpur, Uttar Pradesh" },
    });
    fireEvent.change(mobileInput, { target: { value: "9956455678" } });

    fireEvent.click(saveButton);
    //fireEvent.submit(screen.getByRole("form"));

    //wait for the form submission to complete
    await waitFor(() => {
      expect(onSaveMock).toHaveBeenCalledWith({
        name: "aman",
        email: "aman@gmail.com",
        pincode: "209727",
        address: "Kanpur, Uttar Pradesh",
        mobile: "9956455678",
      });
    });
  });
});

describe("Profile component", () => {
  test("Should show Loading icon", () => {
    render(
      <UserContextProvider>
        <BrowserRouter>
          <Profile />
        </BrowserRouter>
      </UserContextProvider>
    );
    //screen.debug();
    const loader = screen.getByTestId("loading-spinner");
    expect(loader).toBeInTheDocument();
  });

  test("Should show user Details", async () => {
    const mockValue = {
      name: "aman",
      email: "aman@gmail.com",
      pincode: "209727",
      address: "Kanpur, Uttar Pradesh",
      mobile: "9956455678",
    };

    // Use mockResolvedValueOnce instead of mockReturnValueOnce
    axios.get.mockResolvedValueOnce({ data: mockValue });
    render(
      <UserContextProvider>
        <BrowserRouter>
          <Profile />
        </BrowserRouter>
      </UserContextProvider>
    );
    //screen.debug();
    await waitFor(() => {
      const name = screen.getByText("Something went wrong!");
      expect(name).toBeInTheDocument();
    });
  });
});

test("Snapshot Testing", () => {
  const { container } = render(
    <UserContextProvider>
      <BrowserRouter>
        <EditProfile userData={userData} onSave={onSaveMock} />
      </BrowserRouter>
    </UserContextProvider>
  );
  expect(container.innerHTML).toMatchSnapshot();
});
