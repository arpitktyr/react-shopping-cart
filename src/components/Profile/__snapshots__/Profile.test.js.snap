// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing 1`] = `"<form name=\\"profileForm\\" style=\\"width: 100%; display: flex; flex-direction: column;\\"><div class=\\"form-group\\"><label for=\\"name\\">Name</label><input class=\\"form-control\\" type=\\"text\\" id=\\"name\\" name=\\"name\\" aria-invalid=\\"false\\"></div><div class=\\"form-group\\"><label for=\\"email\\">Email</label><input class=\\"form-control\\" type=\\"email\\" id=\\"email\\" name=\\"email\\" aria-invalid=\\"false\\"></div><div class=\\"form-group\\"><label for=\\"pincode\\">Pincode</label><input class=\\"form-control\\" type=\\"text\\" id=\\"pincode\\" name=\\"pincode\\" aria-invalid=\\"false\\"> </div><div class=\\"form-group\\"><label for=\\"address\\">Address</label><textarea class=\\"form-control\\" id=\\"address\\" name=\\"address\\" aria-invalid=\\"false\\"></textarea></div><div class=\\"form-group\\"><label for=\\"mobile\\">Mobile</label><input class=\\"form-control\\" type=\\"text\\" id=\\"mobile\\" name=\\"mobile\\" aria-invalid=\\"false\\"></div><button name=\\"save\\" class=\\"btn btn-success\\" type=\\"submit\\">Save</button></form>"`;
